src/Features/VisualBasic/Portable/ConvertLinq/VisualBasicConvertLinqQueryToLinqMethodProvider.vb

```diff
@@ -0,0 +1,234 @@
+                    Dim fieldInitializers = objectCreation.Initializers.
+                        Select(Function(initializer) SyntaxFactory.InferredFieldInitializer(
+                        SyntaxFactory.IdentifierName(BeautifyName(DirectCast(initializer, IParameterReferenceOperation).Parameter.Name))))
+                    Dim anonymousObjectCreation = SyntaxFactory.AnonymousObjectCreationExpression(
+                        SyntaxFactory.ObjectMemberInitializer(SyntaxFactory.SeparatedList(Of FieldInitializerSyntax)(fieldInitializers)))
+                    secondLambda = CreateLambdaExpression(secondArgumentAnonymousFunction, anonymousObjectCreation)
+                Else
+                    secondLambda = CreateLambdaExpression(secondArgumentAnonymousFunction)
+                End If
+
+                Return CreateInvocationExpression(expression, NameOf(Enumerable.SelectMany), {firstLambda, secondLambda})
+            End Function
+
+            Private Function CreateLambdaExpression(anonymousFunction As IAnonymousFunctionOperation, Optional body As VisualBasicSyntaxNode = Nothing) As LambdaExpressionSyntax
+                If body Is Nothing Then
+                    body = MakeIdentifierNameReplacements(anonymousFunction.Body.Operations.First().Syntax)
+                End If
+                Dim parameters = anonymousFunction.Symbol.Parameters.
+                    Select(Function(p) SyntaxFactory.Parameter(SyntaxFactory.ModifiedIdentifier(BeautifyName(p.Name))))
+                Dim parameterList = SyntaxFactory.ParameterList(SyntaxFactory.SeparatedList(parameters))
+                Dim header = SyntaxFactory.LambdaHeader(
+                    SyntaxKind.FunctionLambdaHeader,
+                    attributeLists:=Nothing,
+                    modifiers:=Nothing,
+                    SyntaxFactory.Token(SyntaxKind.FunctionKeyword),
+                    parameterList,
+                    asClause:=Nothing)
+                Return SyntaxFactory.SingleLineFunctionLambdaExpression(header, body)
+            End Function
+
+            Private Function ProcessSelectClause(expression As ExpressionSyntax, selectClause As SelectClauseSyntax) As ExpressionSyntax
+                Dim selectExpression = selectClause.Variables.First().Expression
+                ' Avoid trivial Select(Funciton(x) x)
+                Dim variables = selectClause.Variables
+                Dim variable = variables.First()
+                Dim bodyExpression = variable.Expression
+
+                If bodyExpression.Kind() = SyntaxKind.IdentifierName Then
+                    Dim identifierName = DirectCast(bodyExpression, IdentifierNameSyntax)
+                    Dim identifierNames = GetIdentifierNames(_semanticModel, identifierName, _cancellationToken)
+                    If identifierNames.IsDefault OrElse identifierNames.Length = 1 Then
+                        Return expression
```

Comment:
>Return expression [](start = 24, length = 17)

The logic here feels very fragile. We don't even check what name did we get back from GetIdentifierNames. And, even if we got the expected name, where is a guarantee that there is only one variable in the picture?